### PR TITLE
Optimize

### PR DESCRIPTION
**Pull request by chaoticlonghair**
For #10.

Don't squash commits when merge into master.
Since each commit in this PR has lots of modifications, several independent commits can make `git bisect` more easily than a big commit.

### CODE REVIEW

File: src/sm2/field.rs [hunk 1]
```diff
@@ -326,82 +339,160 @@ fn raw_mul(a: &FieldElem, b: &FieldElem) -> [u32; 16] {
...
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        let lhs = self.value;
+        let rhs = other.value;
+        if lhs[0] != rhs[0] {
+            return false;
+        }
+        if lhs[1] != rhs[1] {
+            return false;
+        }
```
Comment (u2) [inline src/sm2/field.rs]:
```
lhs[0] == rhs[0] && lhs[1] == rhs[1] ... 
```

File: src/sm2/field.rs [hunk 2]
```diff
@@ -326,82 +339,160 @@ fn raw_mul(a: &FieldElem, b: &FieldElem) -> [u32; 16] {
...
+    }
+}
+
+impl ::std::cmp::Ord for FieldElem {
+    #[inline]
+    fn cmp(&self, other: &Self) -> ::std::cmp::Ordering {
+        let lhs = self.value;
+        let rhs = other.value;
+        if lhs[0] != rhs[0] {
+            return if lhs[0] > rhs[0] {
```
Comment (u2) [inline src/sm2/field.rs]:
```
return lhs[0].cmp(rhs[0])
```

File: src/sm2/field.rs (hunk 1 above)
Comment (chaoticlonghair) [reply src/sm2/field.rs]: OK, I will update this in the last commit then `push --force`.

I just kept the original logic of this piece of code, I did not change it.
I think changes for style and changes for logic should be in different commits.
https://github.com/cryptape/libsm/blob/3f62eb92758df72eb8f5deec71c9702b82d22eaf/src/sm2/field.rs#L395-L405

File: src/sm2/field.rs (hunk 2 above)
Comment (chaoticlonghair) [reply src/sm2/field.rs]: OK, I will update this.

https://github.com/cryptape/libsm/blob/3f62eb92758df72eb8f5deec71c9702b82d22eaf/src/sm2/field.rs#L381-L393

File: src/sm2/field.rs [hunk 3]
```diff
@@ -263,7 +263,7 @@ impl Default for FieldCtx {
     }
 }
 
-#[derive(Copy, Clone)]
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
```
Comment (u2) [inline src/sm2/field.rs:266]: cool

**Review by u2 — APPROVED** [review]

**Pull request merged**